### PR TITLE
Add TapeRulesStatus to the alarm message && Decrease alarm severity.

### LINKED ISSUE #11456: Many workflows stuck in the system unarchived.
**Impact of the bug**
MSRuleCleaner

**Describe the bug**

We are not yet convinced this is an actual bug. It can easily turn out to be a consequence of human activities or an effect caused by misbehavior of some other components not MSRuleCleaner itself. 

After enabling the alarms in `MSRuleCleaner` for workflows sitting in announced  and not archived for more than 30 days, we've found ourselves in a situation with **1157** workflows violating the above threshold [1](https://github.com/dmwm/WMCore/issues/11396#issuecomment-1403703561). Some of them sitting in the system since 2021. All of them with one and the same reason: `Not properly cleaned workflow`

Here is the full list  of workflow names and dates since sitting in `announced` and not being `archived`.

The current issue is to track the operational issues related to investigation of those workflows and the eventual action items to be taken in order to push them out of the system. 

**How to reproduce it**
Not clear yet.

**Expected behavior**
Not to leave so many workflows for a long period of time stuck in the system.

**Additional context and error message**
[1]
https://github.com/dmwm/WMCore/issues/11396#issuecomment-1403703561

### PR DESCRIPTION
**Pull request by todor-ivanov**
Fixes #11456 

#### Status
ready

#### Description
With the current PR we are decreasing the alarm severity for all workflows stuck in the system unarchived for more than a month and we also add the list of `TapeRulesStatus` and the flag `ParentageResolved` to the alarm message, because those are the two most probable reasons for getting a workflow stuck into the system.   

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None 

#### External dependencies / deployment changes
No

### CODE REVIEW

**Comment by todor-ivanov** [comment]
One thing must be mentioned here: 
While working on the tests of this PR I found out that, because we decided to send alarms with more clear messages as early as possible (basically from the `_dispatcher` method) and also from the archival pipeline as well, we obviously have ended up raising one and the same alarm per every workflow twice: [1]. Instead of creating a separate bug issue just for that. I am fixing this with my next commit to the current PR.  

FYI @amaltaro 

[1]
```
In [88]: msr._dispatchWflow(msrWflows['pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143'])
2023-04-03 14:10:36,333:INFO:MSRuleCleaner: Tape rule: a92d525de4f3464d9c599564f68de11a in non final state: SUSPENDED for workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143
2023-04-03 14:10:36,334:INFO:MSRuleCleaner: Skipping workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143 - tape transfers are not yet completed. Will retry again in the next cycle.
2023-04-03 14:10:36,340:CRITICAL:MSRuleCleaner: Found a workflow not archived since Sat Jan 28 00:39:48 2023.
Workflow: 
{'ParentageResolved': True,
 'RequestName': 'pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143',
 'TapeRulesStatus': [('a92d525de4f3464d9c599564f68de11a',
                      'SUSPENDED',
                      '/NMSSM_XToYHTo2B2G_MX-320_MY-100_TuneCP5_13TeV-madgraph-pythia8/RunIISummer16MiniAODv3-PUMoriond17_94X_mcRun2_asymptotic_v3-v1/MINIAODSIM'),
                     ('a92d525de4f3464d9c599564f68de11a',
                      'SUSPENDED',
                      '/NMSSM_XToYHTo2B2G_MX-320_MY-100_TuneCP5_13TeV-madgraph-pythia8/RunIISummer16MiniAODv3-PUMoriond17_94X_mcRun2_asymptotic_v3-v1/MINIAODSIM')],
 'TransferDone': True,
 'TransferTape': False}
Has exceeded the Status Advance Timeout of: 720 hours for its current status: announced.
Additional info: Skipping workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143 - tape transfers are not yet completed. Will retry again in the next cycle.
Actions need to be taken!
2023-04-03 14:10:36,369:CRITICAL:MSRuleCleaner: Found a workflow not archived since Sat Jan 28 00:39:48 2023.
Workflow: 
{'ParentageResolved': True,
 'RequestName': 'pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143',
 'TapeRulesStatus': [('a92d525de4f3464d9c599564f68de11a',
                      'SUSPENDED',
                      '/NMSSM_XToYHTo2B2G_MX-320_MY-100_TuneCP5_13TeV-madgraph-pythia8/RunIISummer16MiniAODv3-PUMoriond17_94X_mcRun2_asymptotic_v3-v1/MINIAODSIM'),
                     ('a92d525de4f3464d9c599564f68de11a',
                      'SUSPENDED',
                      '/NMSSM_XToYHTo2B2G_MX-320_MY-100_TuneCP5_13TeV-madgraph-pythia8/RunIISummer16MiniAODv3-PUMoriond17_94X_mcRun2_asymptotic_v3-v1/MINIAODSIM')],
 'TransferDone': True,
 'TransferTape': False}
Has exceeded the Status Advance Timeout of: 720 hours for its current status: announced.
Additional info: Not properly cleaned workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143
Actions need to be taken!
2023-04-03 14:10:36,378:INFO:MSRuleCleaner: plineArchive: Proper conditions not met: <@========== WMException Start ==========@>
Exception Class: MSRuleCleanerArchivalSkip
Message: Not properly cleaned workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143
	ClassName : None
	ModuleName : WMCore.MicroService.MSRuleCleaner.MSRuleCleaner
	MethodName : __init__
	ClassInstance : None
	FileName : /data/WMCore.venv3/srv/WMCore/src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
	LineNumber : 66
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>. Skipping archival in the current cycle.

```

**Comment by todor-ivanov** [comment]
And here is the result of my final fix: [1] 
No alarm repetitions at all && every single reason contributing to the archival delay is properly logged into the alarm. 

[1]
```
In [133]: msr._dispatchWflow(msrWflows['pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143'])
2023-04-03 15:01:00,016:INFO:MSRuleCleaner: Tape rule: a92d525de4f3464d9c599564f68de11a in non final state: SUSPENDED for workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143
2023-04-03 15:01:00,016:INFO:MSRuleCleaner: Skipping workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143 - tape transfers are not yet completed. Will retry again in the next cycle.
2023-04-03 15:01:00,048:CRITICAL:MSRuleCleaner: Found a workflow not archived since Sat Jan 28 00:39:48 2023.
Workflow: 
{'ParentageResolved': True,
 'RequestName': 'pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143',
 'TapeRulesStatus': [('a92d525de4f3464d9c599564f68de11a',
                      'SUSPENDED',
                      '/NMSSM_XToYHTo2B2G_MX-320_MY-100_TuneCP5_13TeV-madgraph-pythia8/RunIISummer16MiniAODv3-PUMoriond17_94X_mcRun2_asymptotic_v3-v1/MINIAODSIM')],
 'TransferDone': True,
 'TransferTape': False}
Has exceeded the Status Advance Timeout of: 720 hours for its current status: announced.
Additional info: Skipping workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143 - tape transfers are not yet completed. Will retry again in the next cycle.
Not properly cleaned workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143

Actions need to be taken!
2023-04-03 15:01:00,056:INFO:MSRuleCleaner: plineArchive: Proper conditions not met: <@========== WMException Start ==========@>
Exception Class: MSRuleCleanerArchivalSkip
Message: Not properly cleaned workflow: pdmvserv_task_HIG-RunIISummer15wmLHEGS-05610__v1_T_220914_155035_4143
	ClassName : None
	ModuleName : WMCore.MicroService.MSRuleCleaner.MSRuleCleaner
	MethodName : __init__
	ClassInstance : None
	FileName : /data/WMCore.venv3/srv/WMCore/src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
	LineNumber : 66
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>. Skipping archival in the current cycle.

```

**Comment by todor-ivanov** [comment]
Hi @vkuznet, 

While discussing about redirecting those  alarms to DM Team, I have two inquiries to make to you (hope the answer to those  won't take much of your time):
*   Do we have a well defined reposirotry to which we must create a PR for defining the new rerouting rules for those alarms
* Can we start aggregating those alarms in the alert manager so that people (DM and Us) receive only a single message/alarm containing all of those workflows/tape rule Ids etc.

FYI: @amaltaro

**Comment by vkuznet** [comment]
Todor, CMS Monitoring configs are available in [gitlab repo](https://gitlab.cern.ch/cmsmonitoring/cmsmon-configs) and what you're looking for is [AlertManager config](https://gitlab.cern.ch/cmsmonitoring/cmsmon-configs/-/blob/master/alertmanager/alertmanager.yaml) where you can provide regular PR, of course I would assume you'll learn [AM configuration](https://prometheus.io/docs/alerting/latest/configuration/) to understand the logic.

Regarding alerts aggregation. AlertManager by itself does not provide aggregation, even though its routes allow alerts flow. Said that, in CMS Monitoring we developed [intelligence service](https://github.com/dmwm/CMSMonitoring/tree/master/src/go/intelligence) which does the aggregation. So far the service consumes GGUS, and SNOW tickets to annotate our dashboards, but its logic and architecture will allow alert aggregation. Said that, to make aggregation someone needs to clearly define set of requirements and process of aggregation, then this can be easily implemented in intelligence service. Basically it allows to consume alerts from multiple sources and yield alerts since it is based on AM APIs. But, this will be separate development and with Ceyhun departure I'm not aware of anyone who can manage this task (I can do it but my time is tight to WMCore issues rather then monitoring one).

**Comment by vkuznet** [comment]
Todor, on further reading about AM it may be possible to do some aggregation if alerts in questions will have common attributes. I found one recipe which we may explore, see [here](https://stackoverflow.com/questions/68770055/alertmanager-to-send-aggregated-consolidated-alert-to-webhook). Therefore, it will require properly setup alert attributes among those which you may want to aggregate. For instance, if a specific agent will issue multiple alerts in some time interval those can be grouped by agent name, and only one alert will be issue to a specific web hook.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py [hunk 1]
```diff
@@ -216,7 +219,8 @@ def docSchema(self):
             ('RequestTransition', [], list),
             ('IncludeParents', False, bool),
             ('InputDataset', None, (bytes, str)),
-            ('ParentDataset', None, (bytes, str))]
+            ('ParentDataset', None, (bytes, str)),
+            ('StatusAdvanceExpiredMsg', "", (bytes, str))]
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py:223]: Even though it's not going to be consistent with the rest of the schema, can you please decide which data type it's supposed to be (now that we are fully in python3).

**Review by amaltaro — COMMENTED** [review]
@todor-ivanov I didn't spot any issues with this PR, but I left a few comments/questions along the code for your consideration.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py [hunk 2]
```diff
@@ -412,21 +412,26 @@ def _getLastStatusTransitionTime(self, wflow):
...
+    def _checkStatusAdvanceExpired(self, wflow, additionalInfo=None):
         """
         A method to check if a given status transition has timed out e.g. before
         an alarm being set. The timeout should be configurable and hence taken
-        from the instance attributes.
-        :param wflow:  A MSRuleCleanerWorkflow instance
-        :return:       Bool: True if status alarm time has expired False otherwise
+        from the instance attributes. Preserve any additional info inside the wflow object.
+        :param wflow:          A MSRuleCleanerWorkflow instance
+        :param additionalInfo: Additional info to be preserved within the workflow for later use by the alarm itself. (Default: None)
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:421]: good point

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py [hunk 3]
```diff
@@ -412,21 +412,26 @@ def _getLastStatusTransitionTime(self, wflow):
             self.logger.error("Missing or broken status transition history for %s", wflow['RequestName'])
             return None
 
-    def _isStatusAdvanceExpired(self, wflow):
+    def _checkStatusAdvanceExpired(self, wflow, additionalInfo=None):
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:415]: good point

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py [hunk 4]
```diff
@@ -771,25 +783,30 @@ def getRequestRecords(self, reqStatus):
...
+    def alertStatusAdvanceExpired(self, wflow):
         """
         Send alert notifying that a workflow has spend too much time in a given status
         :param wflow: MSRuleCleanerWorkflow instance
         :additionalInfo: String with additional information.
         :return: none
         """
         # Check if alarm threshold has expired:
-        if not self._isStatusAdvanceExpired(wflow):
+        if not self._checkStatusAdvanceExpired(wflow):
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:794]: Must be a leftover from previous implementation - I was intending to remve the internal call to `_checkStatusAdvanceExpired`, but I must have forgotten. Will do.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py [hunk 5]
```diff
@@ -182,6 +182,8 @@ def docSchema(self):
                               'plineAgentBlock': False},
             "TransferDone": False  # information - returned by the MSOutput REST call.
             "TransferTape": False  # information - fetched by Rucio about tape rules completion
+            "TapeRulesStats": [('36805b823062415c8ee60300b0e60378', 'OK', '/AToZHToLLTTbar_MA-1900_MH-1200_TuneCP5_13TeV-amcatnlo-pythia8/RunIISummer20UL16RECO-106X_mcRun2_asymptotic_v13-v2/AODSIM'),
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py:185]: should read `TapeRulesStatus` - will fix
Hope I've got which typo you were referring to. If not please be more specific.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py (hunk 1 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py:223]: agreed

**Comment by todor-ivanov** [comment]
Thanks for the review @amaltaro 
I think I've addressed all your comments. Please take another look.

**Review by amaltaro — APPROVED** [review]
Looks good to me, thanks Todor.

**Pull request merged**